R/sparrow_integration.R

```diff
@@ -0,0 +1,264 @@
+##' @param gsdb A gene-centric \code{GeneSetDb} object to conform to the relevant peakwise dataset.
+##' @param minsize Minimum number of targets required to consider a geneset valid for analysis.
+##' @param ... Additional arguments to be passed to `ct.prepareAnnotation()`.
+##' @return A new \code{GeneSetDb} object with the features annotated genewise to pathways.
+##' @examples 
+##' data(resultsDF)
+##' data(ann)
+##' gsdb <- ct.GREATdb(ann, gsdb = getMSigGeneSetDb(collection = 'h', species = 'human', id.type = 'entrez'))
+##' show(featureIds(gsdb))
+##' @export
+ct.GREATdb <- function(annotation, 
+                       gsdb = getMSigGeneSetDb(collection = c('h', 'c2'), 
+                                               species = 'human', 
+                                               id.type = 'ensembl'), 
+                       minsize = 10, 
+                       ...){
+  
+  #input check
+  if (!is(gsdb, "GeneSetDb")) {
+    if (is(gsdb, "GeneSetCollection") || is(gsdb, "GeneSet")) {
+      stop("A GeneSetDb is required. GeneSetCollections can be converted to a ", 
+           "GeneSetDb via the GeneSetDb constructor, or a call to ", 
+           "`as(gene_set_collection, 'GeneSetDb')`. See ?GeneSetDb for more ", 
+           "details.")
+    }
+    stop("GeneSetDb required. Please see `?GeneSetDb` for ways to turn your ", 
+         "gene sets into a GeneSetDb, or ")
+  }
+  stopifnot(is(gsdb, 'GeneSetDb'), is(minsize, 'numeric'), (length(minsize) == 1))
+  
+  #check the annotation object
+  annotation <- ct.prepareAnnotation(annotation, ...)
+  
+  #Check to make sure that the geneID column values are present in the provided gsdb
+  genes <-na.omit(unique(annotation$geneID))
+  message(paste0(length(intersect(sparrow::featureIds(gsdb), genes)), ' valid genes found in the supplied gene database (of ', length(genes), ')'))
+
+  #Tabulate targets per gene
+  targetsPerGene <- lapply(genes,
+                         function(x){
+                           return(as.character(unique(annotation$geneSymbol[annotation$geneID %in% x])))
+                         })
+  names(targetsPerGene) <- genes
+  
+  #Pull the lists from the gsdb
+  gsdb_list <- as.list(gsdb)
+  
+  new_gs <- lapply(gsdb_list,
+                   function(x){
+                     return(as.character(unique(unlist(targetsPerGene[x]))))
+                   })
+  new_gs <- new_gs[unlist(lapply(new_gs, length)) >= minsize]
+  message(paste0(length(new_gs), ' adjusted genesets created.'))
+  if(length(new_gs) == 0){stop('Exiting.')}
+  
+  collections <- vapply(names(new_gs), function(x){strsplit(x, split = ';;')[[1]][1]}, character(1))
+  sets <- vapply(names(new_gs), function(x){strsplit(x, split = ';;')[[1]][2]}, character(1))
+  names(new_gs) <- sets
+  
+  out <- lapply(unique(collections),
+                function(z){
+                  new_gs[collections %in% z]
+                })
+  names(out) <- unique(collections)
+  return(sparrow::GeneSetDb(out, ...))
```

Comment:
Last drive-by comment for now: although I'm sure this code does what you want just fine, I'm trying to grok it a bit and it feels like (easily could be wrong) this is re-implementing functionality that I think is already doable through a few already-implemented (and unit tested 😇) higher-level function calls over a GeneSetDb object.

Seems like your fishing through the features of a GeneSetDb object to remove ones that don't exist in your dataset perhaps?

When you `conform(gdb, dataseset, min, min.gs.size = minsize)` the GeneSetDb object to the dataset, it will ignore the features in the `gdb` that aren't in your `dataset`, as well as the genesets that have < 10 matched features.

Everything downstream should "just work", but if you want to drop the features that aren't in your data, as well as the genesets that fail the `minsize` criteria "with prejudice" and just create a GeneSetDb that has only thing things you need, I think you can do all the things you're doing here with something like the following:

(Let's assume the feature_id's in your "annotation" are in the expected places, but let's talk about that live just to be sure)

```r
gdb <- conform(gdb, annotation, min.size = minsize)
gdb.clean <- GeneSetDb(as.data.frame(gdb, active.only = TRUE))
```

The little `n` column you get from the `geneSets(gdb)` also tells you how many features in `gdb` matched to the dataset your `conform()`-ed the `gdb` object to ... you know?

Reply:
This is a good point, but the key is the operations on lines 53-66, which produce a list that needs to be conformed into a GeneSetDb. I vaguely recall that I found that it was much more performant to conform the list prior to creating the GeneSetDB, but I might be mistaken there. I'll try your implementation suggestion and see if it matters.

Reply:
... how _dare_ you, sir ... 🤪

Reply:
I'm sure that it isn't your code- it's that there are a ton of pathways that are very small and so the db -> list -> db switch seems to be slow. 

Another workaround could be to conform the db first to something bigger than minsize, unpack/repack, and conform again...